### DISCUSSION TITLE
Dump voice vectors and enable diarization only.

### DISCUSSION THREAD

**Discussion by steipal** [Purfview/whisper-standalone-win#353] [Ideas]
Great addition with the diarization! :-)
We use diarization as a separate independant step in a complex workflow using FFAStrans. But it would be nice to do it without the extra separate pyannote package now that you have implemented it in faster-whisper-xxl. So in that regard I have two feature requests

1. Fetch speaker embeddings (if available?) and dump. https://github.com/Purfview/whisper-standalone-win/discussions/322#discussioncomment-11457862
2. Make it possible to only do diarization without transcribing.

Keep up the great work! :-)

**Comment by 1001ruchka** [comment]
Please add the ability to diarize by saving the file --diarize_dump but without running transcribing. Now I have to wait for transcribing to finish although I only need the txt file of the dump.

Also from txt dump, with the help of PowerShell script written by ChatGPT, I get srt file with SPEAKER_01, SPEAKER_02, etc. as text.
It would be great if you could do such output in srt in Faster-Whisper-XXL directly without using external scripts.

Then I work with this file in SubtitleEdit and leave only subtitles for the desired speaker in the bilingual video. After that I use [another PowerShell script ](https://github.com/Purfview/whisper-standalone-win/discussions/322#discussioncomment-11182379) and mute other speakers in the video and transcribe the result as usual. In this way I get transcription of bilingual video of acceptable quality.

<details>
  <summary>Script diarize_dump2srt:</summary>
  
```
# Input file path
$inputFile = “video.mp4_diarize_pyannote_v3.1_dump.txt”
# Path to output file
$outputFile = “output.srt”

# Read lines from the file
$content = Get-Content $inputFile

# Initialize subtitle counter
$counter = 1

# Create or clear the output file
Out-File -FilePath $outputFile -Encoding utf8

foreach ($line in $content) {
# Extract start time, end time and text
if ($line -match '\[\s*(\d{2}:\d{2}:\d{2}\.\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2}\.\d{3})\]\s*\w+\s*(SPEAKER_\d+)') {
$startTime = $matches[1]
$endTime = $matches[2]
$text = $matches[3]

# Subtitle generation
$srtEntry = @”
$counter
$startTime --> $endTime
$text

“@
# Write subtitle to file
Add-Content -Path $outputFile -Value $srtEntry

# Increase subtitle counter
$counter++
}
}

Write-Host “Conversion completed. The file is saved as $outputFile”
```

</details>

**Comment by Purfview** [comment]
Both features implemented in r245.2

**Reply by steipal** [reply]
Thanks a lot! Looks like exactly what I wished for :-)

**Reply by Purfview** [reply]
You are welcome. What do you do with those embeddings? If you have a python function comparing/assigning known names of speakers then I can add it too.

**Reply by 1001ruchka** [reply]
> What do you do with those embeddings? 

I wrote earlier. I will convert this to subtitles with speaker names SPEAKER_01, SPEAKER_02. Then I manually delete unnecessary speakers leaving only one. I use a script via ffmpeg to mute the voices of other speakers in the recording and do speech recognition. Then I repeat the same thing with a second speaker. In this way I can produce a high quality diarization.
https://github.com/Purfview/whisper-standalone-win/discussions/353#discussioncomment-11494484
https://github.com/Purfview/whisper-standalone-win/discussions/322#discussioncomment-11182379

**Reply by Purfview** [reply]
No, I meant things with embeddings.

**Reply by Purfview** [reply]
@1001ruchka if your audios are heavily multilingual then you can try this:
`--batched --unmerged --multilingual true`

**Reply by steipal** [reply]
@Purfview We have a database of embeddings and use [cosine similarity](https://medium.com/@techclaw/cosine-similarity-between-two-arrays-for-word-embeddings-c8c1c98811b#:~:text=Cosine%20similarity%20measures%20the%20cosine,and%20%2D1%20indicating%20opposite%20vectors.) to find matching known voices. Our implementation is a bit too in-house to currently fit into this repo. The important thing is to obtain the embeddings, and your implementation works just as expected in this regard. If I could wish from the top shelf, a structured json output for both diarization and embeddings would be preferrable from a programing perspective, but it's not important.

I have however noticed two anomalies (?):
1. Even if I specify --diarize_only it still want to load a whisper model. I could just specify tiny just to get a quick load...but ->
2. ...it seems that the _transcription model_ specified in the first  **diarize only** run, influence a subsequent **transcription only** run. Does that make sense?

Thanks again for your great efforts! :-)

**Reply by Purfview** [reply]
>I have however noticed two anomalies (?):
>1. Even if I specify --diarize_only it still want to load a whisper model. I could just specify tiny just to get a quick load...but ->
>2. ...it seems that the _transcription model_ specified in the first  **diarize only** run, influence a subsequent **transcription only** run. >Does that make sense?

Sorry but that doesn't make sense for me. Describe both more specifically with the commands used.
